src/libponyrt/actor/actor.c

```diff
@@ -230,10 +238,122 @@ static void try_gc(pony_ctx_t* ctx, pony_actor_t* actor)
+    &actor->extra_throttled_msgs_sent, memory_order_relaxed)) *
+    actor_throttlefactor);
+}
+
+// return true if mute occurs
+static bool maybe_mute(pony_actor_t* actor)
+{
+  // if we become muted as a result of handling a message, bail out now.
+  // we aren't set to "muted" at this point. setting to muted during a
+  // a behavior can lead to race conditions that might result in a
+  // deadlock.
+  // Given that actor's are not run when they are muted, then when we
+  // started our batch, batch size would have been greater then 0.
+  // If any of our message sends would result in the actor sending to a
+  // muted/overloaded/under pressure receiver, the actor->mute_map_count
+  // will have been incremented or the actor->extra_throttled_msgs_sent
+  // would have been incremented. Both of these throttle the batch size
+  // to be smaller and smaller until it reaches 0.
+  //
+  // We will then set the actor to "muted" once the batch size gets
+  // throttled to 0. Once set, any actor sending a message to it will
+  // also have its batch size throttled smaller until it is also muted
+  // when its batch size reaches 0 unless said sender is marked
+  // as overloaded.
+  //
+  // The key points here is that:
+  //   1. We can't set the actor to "muted" until after its finished running
+  //   a behavior.
+  //   2. We should bail out from running the actor and return false so that
+  //   it won't be rescheduled.
+  if(throttled_batch_size(actor) == 0)
+  {
+    // there is a race condition that exists where another thread might
+    // change the `mute_map_count` and/or the `extra_throttled_msgs_sent`
+    // since we've read them to calculate our batch size. We assume that
+    // hasn't happened and mute the actor and deal with that edge case
+    // after.
+    ponyint_mute_actor(actor);
+
+    // ensure we should actually be muted and our batch size didn't grow
+    // due to the race condition that exists in relation to multiple
+    // scheduler threads modifying `mute_map_count` and
+    // `extra_throttled_msgs_sent` concurrently
+    if(throttled_batch_size(actor) == 0)
+      return true;
+    else
+    {
+      // unmute because we shouldn't be muted due to our batch size growing
+      ponyint_unmute_actor(actor);
+      return false;
+    }
+  }
+
+  return false;
+}
+
+static bool batch_limit_reached(pony_actor_t* actor, bool polling, size_t batch)
+{
+  if(!has_flag(actor, FLAG_OVERLOADED)
+    && !polling && (batch == PONY_ACTOR_DEFAULT_BATCH))
+  {
+    // If we hit the default batch size, consider this actor to be overloaded
+    // only if we're not polling from C code. This is to ensure throttled
+    // actors are not accidentally marked as overloaded.
+    // Overloaded actors are allowed to send to other overloaded actors
+    // and to muted actors without being muted or throttled themselves.
+    ponyint_actor_setoverloaded(actor);
+  }
+
+  return !has_flag(actor, FLAG_UNSCHEDULED);
```

Comment:
If it's correct that `pony_schedule()` and `pony_unschedule()` are no longer used, then the `FLAG_UNSCHEDULED` flag can be removed entirely?

Reply:
My understanding is that these are for when Pony is used as a C library.

Reply:
@dipinhora is correct. they are still "used" as part of the library for running Pony code from C. 

cc/ @slfritchie